### PR TITLE
Facility breakdowns

### PR DESCRIPTION
**Pull request by harimohanraj89**
**Story card:** https://www.pivotaltracker.com/story/show/172096574

## Because

CVHOs would like a per-facility breakdown of district cohort reports when downloaded.

## This addresses

This PR makes a few major changes to achieve this:
* Updates the data structure of the cohort analytics query
* Flattens the district cohort report into a single row in the CSV download
* Adds additional rows to the download for every facility in the district.

**Note:** The output formatting of the CSV may change a bit.

**Deployment checklist**
* [ ] Clear dashboard caches

### CODE REVIEW

File: app/views/analytics/districts/show.csv.erb [hunk 1]
```diff
@@ -3,55 +3,61 @@
...
+    followed_up,
+    controlled,
+    uncontrolled,
+    defaulted,
+    "#{controlled_percent}%",
+    "#{uncontrolled_percent}%",
+    "#{default_percent}%",
+    nil
+  ]
+end).to_csv -%>
```
Comment (harimohanraj89) [inline app/views/analytics/districts/show.csv.erb:60]: Changes up to this point are adding per-facility rows for the cohort report

File: app/views/analytics/districts/show.csv.erb [hunk 2]
```diff
@@ -63,7 +69,7 @@
   dates = dates_for_periods(@period, 3, include_current_period: @show_current_period)
   ["Follow up patients", "Registered patients"].each do |metric|
     dates.each do |date|
-      headers << "#{metric} #{date}"
+      headers << "#{metric} #{format_period(@period, date)}"
```
Comment (harimohanraj89) [inline app/views/analytics/districts/show.csv.erb:72]: `Registrations 2020-04-01` -> `Registrations Apr 2020`

File: app/controllers/analytics/districts_controller.rb [hunk 3]
```diff
@@ -82,6 +83,12 @@ def set_dashboard_analytics(period, prev_periods)
       end
   end
 
+  def set_facility_keys
+    facilities = @organization_district.facilities.order(:name).pluck(:id, :name).to_h
+
+    @facility_keys = { total: "Total" }.merge(facilities).with_indifferent_access
+  end
```
Comment (harimohanraj89) [inline app/controllers/analytics/districts_controller.rb:90]: We pre-fetch a list of facilities in alphabetical order to help generate the rows of data in the CSV in order.

File: app/views/analytics/facilities/show.csv.erb [hunk 4]
```diff
@@ -18,11 +18,11 @@
...
-    registered = analytics[:registered]
-    followed_up = analytics[:followed_up]
-    defaulted = analytics[:defaulted]
-    controlled = analytics[:controlled]
-    uncontrolled = analytics[:uncontrolled]
+    registered = analytics.dig(:registered, :total) || analytics[:registered]
+    followed_up = analytics.dig(:followed_up, :total) || analytics[:followed_up]
+    defaulted = analytics.dig(:defaulted, :total) || analytics[:defaulted]
+    controlled = analytics.dig(:controlled, :total) || analytics[:controlled]
+    uncontrolled = analytics.dig(:uncontrolled, :total) || analytics[:uncontrolled]
```
Comment (harimohanraj89) [inline app/views/analytics/facilities/show.csv.erb:25]: We leave the old `analytics[:registered]` method in here to ensure that the code still works with the older data structure that'll be in cache when we deploy this change. 48 hours after we deploy this change, we can clean up the second clause.

File: app/queries/cohort_analytics_query.rb [hunk 5]
```diff
@@ -41,13 +38,24 @@ def patient_counts(cohort_start, cohort_end, report_start, report_end)
...
-      defaulted: registered_patients.size - followed_up_patients.size,
-      controlled: controlled_patients.size,
-      uncontrolled: uncontrolled_patients.size
-    }
+      registered: { total: registered_patients.size, **registered_counts },
+      followed_up: { total: followed_up_patients.size, **followed_up_counts },
+      defaulted: { total: registered_patients.size - followed_up_patients.size, **defaulted_counts },
+      controlled: { total: controlled_patients.size, **controlled_counts },
+      uncontrolled: { total: uncontrolled_patients.size, **uncontrolled_counts }
+    }.with_indifferent_access
```
Comment (kitallis) [inline app/queries/cohort_analytics_query.rb:58]: Consider removing the individual `symbolize_keys` and do a `deep_symbolize_keys` at the very end.

**Comment by kitallis** [comment]
I've only partially reviewed this, will take another pass at it on 30.04.2020.

File: app/queries/cohort_analytics_query.rb [hunk 6]
```diff
@@ -74,6 +82,6 @@ def followed_up(registered_patients, report_start, report_end)
   end
 
   def controlled(patients)
-    patients.select { |p| p.bp_systolic < 140 && p.bp_diastolic < 90 }
+    patients.where('newest_bps.systolic < 140 AND newest_bps.diastolic < 90')
```
Comment (prabhanshuguptagit) [inline app/queries/cohort_analytics_query.rb:85]: 👌

File: app/queries/cohort_analytics_query.rb (hunk 5 above)
Comment (harimohanraj89) [reply app/queries/cohort_analytics_query.rb:58]: I tried making this change, and remembered why it's this way. The splat (`**registered_counts`) requires the keys to be symbols, and I find the splat operator a lot more readable and scalable than merging hashes together.

Comment (kitallis) [reply app/queries/cohort_analytics_query.rb:58]: Fair enough.

**Comment by harimohanraj89** [comment]
The cohort analytics query object accepts a relation of patients to perform its operations on. I think the best spot to apply HTN filtering would be right before we pass a relation of patients into this object (somewhere in [this vicinity](https://github.com/simpledotorg/simple-server/blob/master/app/models/organization_district.rb#L17-L20) to start). That's a single place to make the modification that would reflect on all cohort reporting, both district and facility level.

**Pull request merged**